qutip/lattice.py

```diff
@@ -0,0 +1,814 @@
+        # current definition of x
+        xs = np.diagflat(R)
+        return Qobj(xs)
+
+    def operator_at_cells(self, op, cells):
+        """
+        A function that returns an operator matrix that applies op to specific
+        cells specified in the cells list
+
+        Parameters
+        ----------
+        op : qobj
+            Qobj representing the operator to be applied at certain cells.
+
+        cells: list of int
+            The cells at which the operator op is to be applied.
+
+        Returns
+        -------
+        isherm : bool
+            Returns the new value of isherm property.
+        """
+        if (not isinstance(cells, list)):
+            if (isinstance(cells, int)):
+                cells = [cells]
+            else:
+                raise Exception("cells in operator_at_cells() need to be a \
+                                list of ints.")
+        else:
+            for i in range(len(cells)):
+                if (not isinstance(cells[i], int)):
+                    print("cells[", i, "] is not an int!")
+                    raise Exception("elements of cells is required to be \
+                                    ints.")
+
+        nSb = self._H_intra.shape
+        if (not isinstance(op, qutip.qobj.Qobj)):
+            raise Exception("op in operator_at_cells need to be Qobj's. \n")
+        nSi = op.shape
+        if (nSb != nSi):
+            raise Exception("op in operstor_at_cells() is required to \
+            have the same dimensionality as cell_Hamiltonian.")
+
+        (xx, yy) = np.shape(op)
+        row_ind = np.array([])
+        col_ind = np.array([])
+        data = np.array([])
+        NS = self._length_of_unit_cell
+        nx_units = self.num_cell
+        ny_units = 1
+        for i in range(nx_units):
+            for j in range(ny_units):
+                lin_RI = i + j * nx_units
+                if ((i in cells) and j == 0):
+                    for k in range(xx):
+                        for l in range(yy):
+                            row_ind = np.append(row_ind, [lin_RI*NS+k])
+                            col_ind = np.append(col_ind, [lin_RI*NS+l])
+                            data = np.append(data, [op[k, l]])
+                else:
+                    for k in range(xx):
+                        row_ind = np.append(row_ind, [lin_RI*NS+k])
+                        col_ind = np.append(col_ind, [lin_RI*NS+k])
+                        data = np.append(data, [1])
+
+        M = nx_units*ny_units*NS
+        op_H = csr_matrix((data, (row_ind, col_ind)), [M, M], dtype=np.complex)
+        return Qobj(op_H)
+
+    def plot_dispersion(self):
+        """
+        Plots the dispersion relationship for the lattice with the specified
+        number of unit cells. The dispersion of the infinte crystal is also
+        plotted if num_cell is smaller than MAXc.
+        """
+        MAXc = 20     # Cell numbers above which we do not plot the infinite
```

Comment:
What is the reason for choosing MAXc=20? It might be a default parameter of the method, set equal to 20.

Reply:
The plot_dispersion() function plots the dispersion of the lattice over the dispersion of the infinite crystal, even if the user does not ask for it. But for 20-ish number of cells, the dispersion of the crystal pretty much starts to look like the infinite crystal dispersion itself.